### ISSUE TITLE
Implement bachnorm layer and add it to some tests

### ISSUE THREAD

**Issue by Mikolaj** [Mikolaj/horde-ad#42]
https://en.wikipedia.org/wiki/Batch_normalization
https://penkovsky.com/neural-networks/day4/

I'm sure there are good blog posts about that. Looking this up in pytorch and/or tensorflow API may help, too.

Perhaps adding this to the existing tests with convolutional and recurrent neural networks would be a good idea. We could measure how the training improves. With batchnorm perhaps we could switch activation function to relu, in examples where it's not already relu, and not lose convergence. Probably normalizing data from -1 to 1 (zero mean, unit variance) would also be useful for relu, so a convenient tools for normalizing is a related task (probably batchnorm is just much more general, e.g., it needs to be differentiable).

Two other keywords that may be related or may be even the same thing as one of the above, so worth googling:

* regularization functions, to keep params not too small or big

* activity_regularizer: Regularizer function applied to the output of the layer (its "activation")

**Comment by Mikolaj** [comment]
A TODO left over from an experimental implementation of the algorithms in https://arxiv.org/pdf/2202.08587.pdf mentions similar kewords. When this ticket is closed, let's create a new one that mentions the parts of the TODO not completed and clarifies the terms used.

> Consider weight decay instead of normalization/regularization/gradient clipping. Verify if it's on by default in pytorch or tensorflow. When the forward gradient paper is published with source code, verify what they have used to avoid gradient explosion. If nothing, investigate why I'm quikly getting NaN in mnistTestCase2D 1024 1024 test and they don't.

**Comment by Mikolaj** [comment]
Related transcript from datahaskell/lobby (https://matrix.to/#/!JSHTXUPHaIJhjEINXX:matrix.org?via=m.topoi.dev&via=gitter.im&via=matrix.org) follows the line:

---

Man of Letters

Hi! I'm implementing yet another automatic differentiation library with a bunch of co-conspirators and a major struggle in that is wrapping my head around some data science concepts, e.g., data normalization (https://github.com/Mikolaj/horde-ad/issues/42).

anton_5

Have you seen this paper https://arxiv.org/pdf/1502.03167.pdf ? It’s basically Algorithm 1 on page 3 and Algorithm 2 on page 4.
Man of Letters

Man of Letters

Oh, thanks, well explained, definitely easier to follow than source code of tensorflow or pytorch --- BTW, is that exactly what they are implementing as batch normalization? Things could have changed since 2015...

anton_5

From the TF docs https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization it appears to me that it’s implemented exactly as discussed in the aforementioned paper. See “During training” and “During inference” sections.

**Comment by Mikolaj** [comment]
Though maybe normalization is not essential after all: https://arxiv.org/pdf/2503.10622
